File: framework/fit/python/fit_flowable/emitter/default_emitter.py
# -- encoding: utf-8 --
# Copyright (c) Huawei Technologies Co., Ltd. 2024-2024. All rights reserved.
"""
功 能：Emitter 的可观测实现。
"""
from threading import Lock
from typing import List

from fit_flowable.core.emitter import Emitter


class DefaultEmitter(Emitter):
    """
    Emitter 的可观测的实现。
    """

    def __init__(self):
        self._observers: List[Emitter.Observer] = []
        self._lock: Lock = Lock()

    def emit(self, data) -> None:
        observer_list = self._get_observers()
        for observer in observer_list:
            observer.on_data_emitted(data)

    def complete(self) -> None:
        observer_list = self._get_observers()
        for observer in observer_list:
            observer.on_completed()

    def fail(self, cause: Exception) -> None:
        observer_list = self._get_observers()
        for observer in observer_list:
            observer.on_failed(cause)

    def observe(self, observer: Emitter.Observer) -> None:
        if observer is None:
            return
        with self._lock:
            self._observers.append(observer)

    def _get_observers(self) -> List[Emitter.Observer]:
        with self._lock:
            return self._observers.copy()

File: framework/fit/java/fit-reactor/src/main/java/com/huawei/fitframework/flowable/emitter/DefaultEmitter.java
/*
 * Copyright (c) Huawei Technologies Co., Ltd. 2024-2024. All rights reserved.
 */

package com.huawei.fitframework.flowable.emitter;

import com.huawei.fitframework.flowable.Emitter;
import com.huawei.fitframework.util.LockUtils;

import java.util.ArrayList;
import java.util.List;

/**
 * 表示 {@link Emitter} 的可观测的实现。
 *
 * @param <T> 表示发送数据的类型的 {@link T}。
 * @author 季聿阶
 * @since 2024-02-13
 */
public class DefaultEmitter<T> implements Emitter<T> {
    private final List<Observer<T>> observers = new ArrayList<>();
    private final Object lock = LockUtils.newSynchronizedLock();

    @Override
    public void emit(T data) {
        List<Observer<T>> observerList = this.getObservers();
        for (Observer<T> observer : observerList) {
            observer.onEmittedData(data);
        }
    }

    @Override
    public void complete() {
        List<Observer<T>> observerList = this.getObservers();
        for (Observer<T> observer : observerList) {
            observer.onCompleted();
        }
    }

    @Override
    public void fail(Exception cause) {
        List<Observer<T>> observerList = this.getObservers();
        for (Observer<T> observer : observerList) {
            observer.onFailed(cause);
        }
    }

    private List<Observer<T>> getObservers() {
        List<Observer<T>> copied;
        synchronized (this.lock) {
            copied = new ArrayList<>(this.observers);
        }
        return copied;
    }

    @Override
    public void observe(Observer<T> observer) {
        if (observer != null) {
            synchronized (this.lock) {
                this.observers.add(observer);
            }
        }
    }
}
